### PR TITLE
Refactor template deployment cmdlets.

### PR DESCRIPTION
**Pull request by Tiano2017**
## Checklist

- [ ] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [ ] The title of the PR is clear and informative
- [ ] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [ ] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

**Comment by isra-fel** [comment]
For PR big like this, [a cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr) is required.

As for the online-version, you can add them manually.

**Comment by Tiano2017** [comment]
> For PR big like this, [a cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr) is required.
> 
> As for the online-version, you can add them manually.

Created an issue here: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/462

File: src/Resources/ResourceManager/Implementation/Deployments/GetAzureSubscriptionDeploymentCmdlet.cs [hunk 1]
```diff
@@ -37,25 +38,25 @@ public class GetAzureDeploymentCmdlet : ResourceManagerCmdletBase
...
         [Alias("DeploymentName")]
-        [Parameter(Position = 0, ParameterSetName = GetAzureDeploymentCmdlet.DeploymentNameParameterSet, Mandatory = false,
+        [Parameter(Position = 0, ParameterSetName = GetAzureSubscriptionDeploymentCmdlet.DeploymentNameParameterSet, Mandatory = false,
             HelpMessage = "The name of deployment.")]
         [ValidateNotNullOrEmpty]
         public string Name { get; set; }
 
         [Alias("DeploymentId", "ResourceId")]
-        [Parameter(ParameterSetName = GetAzureDeploymentCmdlet.DeploymentIdParameterSet, Mandatory = false,
+        [Parameter(ParameterSetName = GetAzureSubscriptionDeploymentCmdlet.DeploymentIdParameterSet, Mandatory = true,
```
Comment (isra-fel) [inline src/Resources/ResourceManager/Implementation/Deployments/GetAzureSubscriptionDeploymentCmdlet.cs]: Changing an optional parameter to mandatory is a breaking change, and it cannot be checked in until next breaking change release which is ~May.
I agree that it should be mandatory, but my suggestion is to add a breaking change attribute and make the change later.

File: src/Resources/Resources/help/Get-AzManagementGroupDeploymentOperation.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+## EXAMPLES
+
+### Example 1: Get deployment operations given a deployment name
+```
+PS C:\>Get-AzManagementGroupDeploymentOperation -ManagementGroupId myMG -DeploymentName Deploy01
+```
+
+Gets deployment operations with name "Deploy01" at the management group "myMG".
+
+### Example 2: Get a deployment and get its deployment operations
```
Comment (isra-fel) [inline src/Resources/Resources/help/Get-AzManagementGroupDeploymentOperation.md]: Nice examples!

File: src/Resources/ResourceManager/Implementation/Deployments/GetAzureSubscriptionDeploymentCmdlet.cs (hunk 1 above)
Comment (isra-fel) [reply src/Resources/ResourceManager/Implementation/Deployments/GetAzureSubscriptionDeploymentCmdlet.cs]: This is also the cause to CI failure

Comment (Tiano2017) [reply src/Resources/ResourceManager/Implementation/Deployments/GetAzureSubscriptionDeploymentCmdlet.cs]: This shouldn't be breaking. "Id" is the only parameter in this parameter set. If you don't specify the "id" parameter, you'll land with the default parameter set which is "GetByDeploymentName".

Comment (Tiano2017) [reply src/Resources/ResourceManager/Implementation/Deployments/GetAzureSubscriptionDeploymentCmdlet.cs]: I can change it back to "mandatory" to false if that's the right thing..

**Comment by Tiano2017** [comment]
@isra-fel I've made the changes. can you help take another look?

File: src/Resources/ResourceManager/Implementation/Deployments/GetAzureSubscriptionDeploymentCmdlet.cs (hunk 1 above)
Comment (isra-fel) [reply src/Resources/ResourceManager/Implementation/Deployments/GetAzureSubscriptionDeploymentCmdlet.cs]: Right.. then it's a false positive. Please read [Debugging StaticAnalysis Errors](https://github.com/Azure/azure-powershell/blob/master/documentation/Debugging-StaticAnalysis-Errors.md#breaking-changes) to suppress the warning.

**Comment by isra-fel** [comment]
> @isra-fel I've made the changes. can you help take another look?

Sorry for the late response. I believe you are right about it not being a breaking change. If you could revert the last commit and suppress the warning I would be able to merge. Thanks

**Comment by Tiano2017** [comment]
@isra-fel so the only issue is the CI run is giving an error for that. Is that an bug in the CI run?

**Comment by isra-fel** [comment]
> @isra-fel so the only issue is the CI run gave an error if I make that change. Is that an bug in the CI run?

Our CI runs static analysis on the code. One of the checks is for breaking changes. The error was that it detected one (potential) breaking change.

According to your comment it did *not* break anything after all, so we can suppress the error by following these instructions: https://github.com/Azure/azure-powershell/blob/master/documentation/Debugging-StaticAnalysis-Errors.md#breaking-changes

**Comment by Tiano2017** [comment]
> > @isra-fel so the only issue is the CI run gave an error if I make that change. Is that an bug in the CI run?
> 
> Our CI runs static analysis on the code. One of the checks is for breaking changes. The error was that it detected one (potential) breaking change.
> 
> According to your comment it did _not_ break anything after all, so we can suppress the error by following these instructions: https://github.com/Azure/azure-powershell/blob/master/documentation/Debugging-StaticAnalysis-Errors.md#breaking-changes

updated. can you help take a look and merge?

**Pull request merged**